File: source/library/Bezier.js
//Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt

function Bezier( cp0,  cp1, cp2, cp3)
{
		// Debug 
		this._cp0		=	cp0 ;
		this._cp1		=	cp1 ;
		this._cp2		=	cp2 ;
		this._cp3		=	cp3 ;
		//^^Debug
		
		this._vCP		=	new Vector2(cp0) ;
	    this._cx		=	3.0 * (cp1.x - cp0.x) ; 
	    this._bx		=	3.0 * (cp2.x - cp1.x) - this._cx ; 
	    this._ax		=	cp3.x - cp0.x - this._cx - this._bx ;
	    this._cy		=	3.0 * (cp1.y - cp0.y) ;
	    this._by		=	3.0 * (cp2.y - cp1.y) - this._cy ;
	    this._ay		=	cp3.y - cp0.y - this._cy - this._by ;
		this._length	=	this._CalcLength() ;	 
		this.className	=	'Bezier' ;

}

Bezier.Create = function(cp0, cp1, cp2, cp3)
{
	return new Bezier(cp0, cp1, cp2, cp3) ;
}

Bezier.prototype = {
		constructor: Bezier,

		
		getLength: function()
		{
			return this._length ;
		},
	
	
		getVectorPosition: function(t)
		{
			var pos =  this.getPositionXY(t) ;
	    	return new Vector2(pos.x,pos.y) ;
		},
	
		getPositionXY: function(t)
		{

	    	var tSquared	=	t * t ;
	    	var tCubed		=	tSquared * t ;
	        
	    	var resultx = (this._ax * tCubed) + (this._bx * tSquared) + (this._cx * t) + this._vCP.x ;
	    	var resulty = (this._ay * tCubed) + (this._by * tSquared) + (this._cy * t) + this._vCP.y ;
	        
	    	return {x:resultx,y:resulty} ;
		},

		// Calculate Srt(x'(t)^2 + y'(t)^2)
		//x'(t),y'(t) is d/dt of x(t), y(t) - (Bezier defn getPositionXY(t))
		GetPrimeT: function(tv)
		{

			var	t			=	tv || 1 ;
	    	var tSquared	=	t * t ;
			// work out dx/dt and dy/dt    

	   		var resultx = (3*this._ax * tSquared) + (2*this._bx * t) + (this._cx) ;
	  		var resulty = (3*this._ay * tSquared) + (2*this._by * t) + (this._cy) ;

	    	return Math.sqrt(resultx*resultx + resulty*resulty)	;
		},
		
		// Length of a Parametric curve x(t),y(t)
		// is Int(a,b) [Sqrt(x'(t)*x'(t) + y'(t)*y'(t))] 

		_CalcLength: function(fft)
		{
			var ft	=	fft	|| 1 ;

			//Integrate using Trapezium rule
			var N	 = 	100 ;
			var n	 =	parseInt(ft*N) ; //Math.floor(ft*N) ;
			var dt =	1/N //0.01 == one hundred bits ;
		
			var sum1 	=	(this.GetPrimeT(0) + this.GetPrimeT(ft))/(2*n) ;
			var sum		=	0 ;
//			for t = dt,ft - dt, dt do
//				sum	=	sum	+	 this.GetPrimeT(t)
//			end
			for (var t = dt; t <= ft - dt; t+=dt)
			{
				sum	=	sum	+	 this.GetPrimeT(t) ;
			}

			return sum/n + sum1 ;
		
		},
		
		render: function(rHelper)
		{
			rHelper.drawCircle(this._cp0.x,this._cp0.y,10) ;
			rHelper.drawCircle(this._cp1.x,this._cp1.y,10) ;
			rHelper.drawCircle(this._cp2.x,this._cp2.y,10) ;
			rHelper.drawCircle(this._cp3.x,this._cp3.y,10) ;
			
			for (var t = 0 ; t <= 1 ; t+=0.01)
			{
				var pos = this.getVectorPosition(t)
				rHelper.drawCircle(pos.x,pos.y,1) ;
			}
		},

		
		toString: function()
		{
				return this.className ;
		}



}

File: lua/Bezier.lua
-- Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt
if not Bezier then

Bezier				=	{}
Bezier.className	=	"Bezier"

	
	function Bezier.Create( cp0,  cp1, cp2, cp3)

		local bezier = {
					p1	=	cp0,
					p2	=	cp1,
					p3	=	cp3,
					p4	=	cp0,
					vCP,cx,bx,ax,cy,by,ay,
					className=Bezier.className,	
				}

		bezier.vCP	=	Vector2.Create(cp0) 
	    bezier.cx	=	3.0 * (cp1.x - cp0.x) 
	    bezier.bx	=	3.0 * (cp2.x - cp1.x) - bezier.cx 
	    bezier.ax	=	cp3.x - cp0.x - bezier.cx - bezier.bx
	    bezier.cy	=	3.0 * (cp1.y - cp0.y)
	    bezier.by	=	3.0 * (cp2.y - cp1.y) - bezier.cy
	    bezier.ay	=	cp3.y - cp0.y - bezier.cy - bezier.by

		setmetatable(bezier,{ __index = Bezier })
		bezier.length	= Bezier._CalcLength(bezier)	 
		return bezier 
		
	end
	
	function Bezier.GetLength(bezier)
		return bezier.length
	end
	
	
	function Bezier.GetPosition( bezier,t)
		local x,y =  Bezier.GetPositionXY(bezier,t)
	    return Vector2.Create(x,y)
	end
	
	function Bezier.GetPositionXY(bezier,t)

	 	local   tSquared, tCubed
	    tSquared = t * t
	    tCubed = tSquared * t
	        
	    local resultx = (bezier.ax * tCubed) + (bezier.bx * tSquared) + (bezier.cx * t) + bezier.vCP.x
	    local resulty = (bezier.ay * tCubed) + (bezier.by * tSquared) + (bezier.cy * t) + bezier.vCP.y
	        
	    return resultx,resulty
	end


	function Bezier.GetPrimeT(bezier,tv)
		local	t	=	tv or 1
	 	local   tSquared
	    tSquared = t * t
	    -- work out dx/dt and dy/dt    
	    local resultx = (3*bezier.ax * tSquared) + (2*bezier.bx * t) + (bezier.cx) 
	    local resulty = (3*bezier.ay * tSquared) + (2*bezier.by * t) + (bezier.cy)
	    return math.sqrt(resultx*resultx + resulty*resulty)	
	end
	
	function Bezier.Cleanup(bezier)
		bezier.vCP = nil
	end
	
	function Bezier._CalcLength(bezier,fft)
	
		local ft	=	fft	or 1

		-- Integrate using Trapezium rule
		local N	 = 	100
		local n	 =	math.floor(ft*N)
		local dt =	1/N -- 0.01 == one hundred bits
		
		local sum1 	=	(Bezier.GetPrimeT(bezier,0) + Bezier.GetPrimeT(bezier,ft))/(2*n)
		local sum	=	0
		for t = dt,ft - dt, dt do
			sum	=	sum	+	 Bezier.GetPrimeT(bezier,t)
		end
		return sum/n + sum1
		
	end
	
-- end of class
end

